### PR TITLE
Feature/owin

### PR DESCRIPTION
**Pull request by haf**
This branch aims to bring OWIN support to Suave.

Current state: programming done as far as the spec is concerned; in need of test cases and more examples to validate implementation.

### CODE REVIEW

**Comment by haf** [comment]
@panesofglass @kolektiv I might have something here. I'd appreciate help to run through the tests.

**Comment by panesofglass** [comment]
I was working in my other branch last night and didn't realize you had picked this branch back up. I'll definitely take a look tonight. Thank you!

File: src/Suave/Owin.fs [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+  IDictionary<string, obj>
+
+type OwinApp =
+  OwinEnvironment -> Async<unit>
+
+type OwinAppFunc =
+  Func<OwinEnvironment, Task>
+
+[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
+module OwinConstants =
```
Comment (panesofglass) [inline src/Suave/Owin.fs]: Is it worth potentially pulling in (either as a dependency or source file via paket) the `Constants` definitions from Freya.Core?

File: src/Suave/Utils/Aether.fs [hunk 2]
```diff
@@ -0,0 +1,223 @@
+﻿module Suave.Utils.Aether
```
Comment (panesofglass) [inline src/Suave/Utils/Aether.fs]: Similar to the above; might it be worth managing this as a source file dependency via paket?

**Comment by panesofglass** [comment]
I had a hard time following the dictionary implementations. Why are there so many? Were lenses not sufficient to support the transition between Suave and OWIN? I could certainly see the need for a dictionary type for the purposes of allowing other frameworks to run on top of the server / host component.

**Comment by haf** [comment]
There are two dictionaries in the PR, one Muuutation (working name) that keeps track to headers' mutations. The other is the Owin dictionary interface. The way I read the spec is that the interface of IDictionary`2 is the dynamic dispatch and for headers, a dictionary needed to be returned and changes tracked.

What would be great now is if someone would help out by adding end-to-end tests into the project, so I can fix the implementation. Especially the coordination between first-write-to-body-stream and seeks inside the request and response body streams, need to be fixed.

**Comment by haf** [comment]
OWIN now works!

``` fsharp
  let composedApp =
    path "/yey-owin"
      >>= setHeader "X-Custom-Before" "Before OWIN"
      >>= owinHelloWorld
      >>= setHeader "X-Custom-After" "After OWIN"
```

screen shot 2015-08-30 at 23 04 09
screen shot 2015-08-30 at 23 07 38

File: src/Suave/Utils/Aether.fs (hunk 2 above)
Comment (haf) [inline src/Suave/Utils/Aether.fs]: Yes, that would be nice. We could lift it in using paket and then change the module name using https://github.com/haf/YoLo/blob/master/README. @kolektiv: are you fine with this?

File: src/Suave/Types.fs [hunk 3]
```diff
@@ -449,15 +449,37 @@ module HttpBinding =
...
+    Aether.idLens <-?> HttpContent.SocketTaskPIso
 
 /// The HttpResult is the structure that you work with to tell Suave how to
 /// send the response. Have a look at the docs for HttpContent for further
 /// details on what is possible.
-type HttpResult =
-  { status  : HttpCode
-    headers : (string * string) list
-    content : HttpContent
+and HttpResult =
```
Comment (haf) [inline src/Suave/Types.fs]: This is now a mutually recursive data type with the above HttpContent.

File: src/Suave/Owin.fs [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+                writePreamble = false
+            }
+      }
+      |> succeed
+
+  [<CompiledName "OfOwinFunc">]
+  let ofOwinFunc (owin : OwinAppFunc) =
+    ofOwin (fun e -> Async.AwaitTask ((owin.Invoke e).ContinueWith<_>(fun _ -> ())))
+
+module OwinServerFactory =
```
Comment (haf) [inline src/Suave/Owin.fs]: It would be great if someone would like to test this out for us.

File: src/Suave/Owin.fs [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+        with get key =
+          Lens.get (Lens.get (owinLensLens key) owinRW) !state
+        and set key value =
+          state := Lens.set (Lens.get (owinLensLens key) owinRW) value !state
+
+      member x.Keys =
+        (owinRW :> IDictionary<_, _>).Keys
+
+      member x.Values =
+        invalidOp "omg the amount of lensing needed"
```
Comment (haf) [inline src/Suave/Owin.fs]: Remaining TODO here.

File: src/Suave/Utils/Aether.fs (hunk 2 above)
Comment (kolektiv) [inline src/Suave/Utils/Aether.fs]: Fine with me, go for it.

File: src/Suave/Owin.fs [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+        // 3.2.3 Other Data
+        OwinConstants.callCancelled, ((fun x -> ct), (fun v x -> x)) <--> untyped // TODO: support cancellation token in HttpRequest signalling aborted request
+        OwinConstants.owinVersion, ((fun x -> "1.3"), (fun v x -> x)) <--> untyped
+
+        // per-request storage
+        "userData", HttpContext.userState_ <--> untyped
+      ]
+
+  type internal OwinDictionary(initialState) =
+    // TODO: support streaming optionally
```
Comment (haf) [inline src/Suave/Owin.fs]: Do we support streaming in the first release? Does someone want to hack something as a continuation of this PR?

File: src/Suave/Owin.fs [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+        OwinConstants.requestBody,          HttpContext.request_ >--> HttpRequest.rawForm_ >--> bytesToStream <--> untyped
+        OwinConstants.requestId,            HttpContext.request_ >--> HttpRequest.trace_ >--> Logging.TraceHeader.traceId_ >--> stringlyTyped uint64 <--> untyped
+        // writeable / value
+        OwinConstants.requestUser,          HttpContext.userState_ >--> claimsPrincipal <--> untyped
+
+        // 3.2.2 Response Data
+        // writeable / value
+        OwinConstants.responseStatusCode,   HttpContext.response_ >--> HttpResult.status_ >--> i2sc <--> untyped
+        // writeable / value
+        OwinConstants.responseReasonPhrase, HttpContext.response_ <--> untyped // TODO: add support for modifying phrasing to Core?
```
Comment (haf) [inline src/Suave/Owin.fs]: Outstanding TODO – support to change the response HTTP-code phrase. (!)

File: src/Suave/Owin.fs [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+      (fun x -> upcast new IO.MemoryStream(x)),
+      (fun v x ->
+        use ms = new IO.MemoryStream()
+        v.CopyTo ms // TO CONSIDER: save the stream instead and read from it when writing data
+        ms.ToArray())
+
+    let claimsPrincipal : Property<Map<string, obj>, ClaimsPrincipal> =
+      (fun x ->
+        x
+        |> Map.tryFind "principal" // TODO: add support for ClaimsPrincipal in Core
```
Comment (haf) [inline src/Suave/Owin.fs]: Outstanding issue we should have a look at: supporting ClaimsPrincipal throughout core.

File: src/Suave/Owin.fs (hunk 1 above)
Comment (haf) [inline src/Suave/Owin.fs]: I don't think that has any value-add.

**Comment by haf** [comment]
TODO: We also need to compare headers with `StringComparer.OrdinalIgnoreCase`.

File: src/Suave/Owin.fs (hunk 4 above)
Comment (panesofglass) [inline src/Suave/Owin.fs]: I will focus on this later tonight.

**Comment by panesofglass** [comment]
@haf, amazing work!

**Comment by ademar** [comment]
Nice! You knock it outta the park :)

**Comment by haf** [comment]
screen shot 2015-08-31 at 01 19 45

**Comment by haf** [comment]
ping @ademar – could you have a look at the last commit? There are some minor breaking changes there.

**Comment by ademar** [comment]
@haf changes look OK to me.

**Comment by panesofglass** [comment]
I will work on some samples during part of my flight tomorrow. Thank you for this!

**Comment by haf** [comment]
@ademar what's is your opinion on adding ClaimsPrincipal? Is it worthy of a property in the main HttpRequest type?

**Comment by ademar** [comment]
Yeah, +1 for ClaimsPrincipal.

On Mon, Sep 14, 2015 at 2:04 AM, Henrik Feldt <notifications@github.com>
wrote:

> @ademar <https://github.com/ademar> what's is your opinion on adding
> ClaimsPrincipal? Is it worthy of a property in the main HttpRequest type?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SuaveIO/suave/pull/142#issuecomment-139971587>.
>



-- 
Ademar Gonzalez
+1-647-891-3606
http://ademar.name
https://github.com/ademar

**Comment by panesofglass** [comment]
Rather than adding samples, I kept working on test cases. See #295.

**Comment by panesofglass** [comment]
@haf any ideas about getting `OrdinalIgnoreCase` to work? I'm about ready to either copy-and-fix `map.fs` from the F# source or write a new one.

**Comment by adamchester** [comment]
@panesofglass could be related to GetHashCode(), as for why your custom equality or comparison methods are not getting called?

IIRC, collections often check that first an optimisation 
https://twitter.com/panesofglass/status/643968944500527104

**Comment by panesofglass** [comment]
@adamchester, it's possible. I didn't check that method while attempting to debug. I have been trying other experiments here: https://github.com/panesofglass/suave/commits/feature/owin-test

**Comment by panesofglass** [comment]
:fireworks:

**Pull request merged**